### PR TITLE
IGNITE-23347 Sql. Fix comparison of character strings

### PR DESCRIPTION
**Pull request by korlov42**
https://issues.apache.org/jira/browse/IGNITE-23347

This patch
- fix problem when implicit cast was added when comparing operands of character types with different length
- remove unnecessary padding during literal creation for CHAR type 
- remove unnecessary padding in CAST to CHAR operation

---------
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/coercion/CharacterStringComparisonExecutionTest.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        try (ClusterWrapper testCluster = testCluster(typePair, dataProvider(typePair))) {
+            testCluster.process("SELECT c1 < c2 FROM t", checkReturnResult());
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource
+    public void comparisonGreat(CharacterStringPair typePair) throws Exception {
+        try (ClusterWrapper testCluster = testCluster(typePair, dataProvider(typePair))) {
+            testCluster.process("SELECT c1 < c2 FROM t", checkReturnResult());
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/coercion/CharacterStringComparisonExecutionTest.java:62]: the same as above

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/fun/SameFamilyOperandTypeChecker.java [hunk 2]
```diff
@@ -55,7 +55,7 @@ public boolean checkOperandTypes(SqlCallBinding callBinding, boolean throwOnFail
 
     @Override
     public Consistency getConsistency() {
-        return Consistency.LEAST_RESTRICTIVE;
+        return Consistency.NONE;
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/fun/SameFamilyOperandTypeChecker.java:58]: seems now it`s identical to super method and can be removed

File: modules/sql-engine/src/integrationTest/sql/types/char/test_char.test [hunk 3]
```diff
@@ -6,26 +6,24 @@
 query T rowsort
 SELECT ' prefix' || c1 || 'trailing ' FROM (VALUES('1 '::CHAR(5)), (' 2'::CHAR(5)), (''::CHAR(5)), (null)) t(c1);
 ----
- prefix 2   trailing 
- prefix1    trailing 
- prefix     trailing 
+ prefix 2trailing 
+ prefix1 trailing
```
Comment (zstan) [inline modules/sql-engine/src/integrationTest/sql/types/char/test_char.test:10]: seems we still bit incompatible here with PG, do we need to add right space padding ?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/coercion/CharacterStringComparisonExecutionTest.java (hunk 1 above)
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/coercion/CharacterStringComparisonExecutionTest.java]: nice catch! Fixed.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/fun/SameFamilyOperandTypeChecker.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/fun/SameFamilyOperandTypeChecker.java:58]: default implementation is not under our control, thus explicit declaration seems to be safer

File: modules/sql-engine/src/integrationTest/sql/types/char/test_char.test (hunk 3 above)
Comment (korlov42) [reply modules/sql-engine/src/integrationTest/sql/types/char/test_char.test:10]: We agreed to consider CHAR type as alias to VARCHAR. This means Ignite will never be compliant with PG when dealing with CHAR type.

Current behavior (with missing right padding) closer to what we agreed.

**Review by zstan — APPROVED** [review]

**Review by xtern — APPROVED** [review]

**Pull request merged**